h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -205,23 +207,30 @@ public void onBuildFailure(final ParallelModelBuilder.ModelBuildFailure modelBui
       } finally {
         parallelSearchGridLock.unlock();
       }
-      attemptBuildNextModel(parallelModelBuilder, null);
+    }
+
+    @Override
+    public void afterBuildProcessing(ParallelModelBuilder parallelModelBuilder, Model finishedModel) {
+      attemptBuildNextModel(parallelModelBuilder, finishedModel);
     }
 
     private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuilder, final Model previousModel) {
-      // Attempt to train next model
-      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
-      if (nextModelParams != null
-              && isThereEnoughTime()
-              && !_job.stop_requested()
-              && !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos())) {
 
-        // FIXME: ModelFeeder's <MP> type parameter is hiding GridSearch's one. Ideally would have moved ModelFeeder outside but it is too coupled with GridSearch class.
-        reconcileMaxRuntime(grid._key, hyperspaceIterator, nextModelParams);
+      synchronized (hyperspaceIterator) {
```

Comment:
I still dont understand why this whole block needs to be synchronized instead of the old approach with null check

Reply:
@honzasterba because old version was not dealing with max_models. 
The problem is that line
```
final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
```
takes next model parameters before all the checks in if statement. So as we are working in a parallel fashion there is a chance for a following scenario:
1) ThreadA initialises nextModelParameters before `if` checks
2) ThreadB in the meantime also takes model parameters and has them processed ( which increases counter of completed models). Let's assume counter was increased to the `max_models` value.
3) ThreadA continues and checking conditions in `if` statement. `hasReachedMaxModels` now will reject this model parameters and we will end up with `max_models - 1` number of built model parameters.

Reply:
ok, sounds reasonable, could we then avoid this synchronized block by moving the check `!parallelModelBuilder.hasReachedMaxModels(hyperspaceIterator.max_models())` to `getNextModelParams` https://github.com/h2oai/h2o-3/pull/4191/files/bd981f568c5ca42afc37f088c4f7a008d3f8ca8a#diff-d12a4b90c0799eec2a3e8439901328e5R260

Reply:
@honzasterba no, we can't do this as well.
Following scenario will lead to a case when we have built more then `max_models` models ( how much more depends on level of parallelisation).

if we do like I think you are suggesting:
```
while (params == null) {
        if (hyperSpaceWalker.hasNext(model) && !parallelModelBuilder.hasReachedMaxModels(hyperspaceIterator.max_models())) { ...
```
we will have a chance to get following scenario:

1) ThreadA passed `hasNext` check. ThreadA passed `hasReachedMaxModels` check
2) ThreadB passed `hasNext` check. ThreadB passed `hasReachedMaxModels` check
3) ThreadA retrieved model parameters, built a model and only now updated counter for completed models. Let's assume that counter now equals to `max_models`. Note that there are still plenty of grid permutations out there.
4) ThreadB ( as it has already passed all the checks before counter's update) retrieved model parameters, built a model. This model will be `max_model + 1`th model.

So we do need synchronisation between  checks and actual retrieval of the parameters. It is similar to the problem with `size()` method in let's say `ConcurrentHashMap` https://stackoverflow.com/questions/10313556/concurrent-hashmap-checking-size. In some concurrent scenarios it is not crucial to check size.. more important to have high concurrent performance. But once you need to control exact size - you have to synchronise write operations ( in our case it is equivalent to building a model).

Reply:
I am looking at the old code and exactly the same situation was possible before, so this is something @Pscheidl did not catch originally?

Reply:
@honzasterba your question is related to the origin of the PR. Early stopping in terms of number of models was embedded inside iterator's hasNext method. Pavel did not have checks for max models outside of the iterator.

Reply:
@honzasterba the difference I think is that before there was only one counter `in progress models`. But in this PR I added completed counter. So there is a chance to have them in an inconsistent state when one was decreased and another was not yet increased.

Reply:
I am sorry, but I do not understand what are you trying to say.

Nevertherless I have gone through the code again and I think I see your point in there being a need for synchronisation here. Are you sure that the synchronized block is as small as it needs to be? I think that at least `reconcileMaxRuntime`  and `parallelModelBuilder.run` do not need to be inside a synchronized block. 

Also it might be more readable to use a ReentrantLock as in other places in parallelmodelbuilder to make it more obvious for what we are locking?

Reply:
@honzasterba I don't think that we can move `reconcileMaxRuntime` and `parallelModelBuilder.run` outside of the synchronisation as they are both dependant on `nextModelParams` variable

> Also it might be more readable to use a ReentrantLock as in other places in parallelmodelbuilder to make it more obvious for what we are locking?

I will try to improve readability

Reply:
but you can call nextModelParams in a synchronized block and then do the rest outside, no?

Reply:
@honzasterba sure, i just don't see how we can synchronise over only `if` conditions and  `getNextModelParams` call. Need to think about it more.

Reply:
something like
```suggestion
     MPF nextModelParams = null;
      synchronized (hyperspaceIterator) {
         if(checks)
           nextModelParams = getNextModelParams()
       }
       if (nextModelParams != null) train
       else noMoreModels() 
```
could work

Reply:
@honzasterba yes, something like this. But it would not work. Because we need this line from `parallelModelBuilder.run` to be synchronised:
```
_modelInProgressCounter.incrementAndGet();
```
Otherwise it is the same as your second suggestion where we end up with `max_models + 1` models. Not sure if it makes sense to increment counter separately from `attemptBuildNextModel` as 1) `reconcileMaxRuntime` should be called before `parallelModelBuilder.run` method and 2) `parallelModelBuilder.run` is a non-blocking training

Reply:
I actually was trying, of course, to reduce scope of synchronisation (before holidays) and ended up with what I have in current version

Reply:
I did not check max models in my original implementation, that's true. I thought walkers are the ones who control max model amount, which I can see is not true right now.

Walker knows how many models it "gave out", would it be possible to have such logic there ?